Core/MIPS/IR/IRPassSimplify.cpp

```diff
@@ -0,0 +1,273 @@
+			} else if (gpr.IsImm(inst.src1)) {
+				gpr.SetImm(inst.dest, gpr.GetImm(inst.src1));
+			} else {
+				gpr.MapDirtyIn(inst.dest, inst.src1);
+				goto doDefault;
+			}
+			break;
+
+		case IROp::Mult:
+		case IROp::MultU:
+			gpr.MapInIn(inst.src1, inst.src2);
+			goto doDefault;
+
+		case IROp::MovZ:
+		case IROp::MovNZ:
+			gpr.MapInInIn(inst.dest, inst.src1, inst.src2);
+			goto doDefault;
+
+		case IROp::Min:
+		case IROp::Max:
+			gpr.MapDirtyInIn(inst.dest, inst.src1, inst.src2);
+			goto doDefault;
+
+		case IROp::Clz:
+		case IROp::BSwap16:
+		case IROp::BSwap32:
+		case IROp::Ext16to32:
+		case IROp::Ext8to32:
+			gpr.MapDirtyIn(inst.dest, inst.src1);
+			goto doDefault;
+
+		case IROp::FMovFromGPR:
+			if (gpr.IsImm(inst.src1)) {
+				out.Write(IROp::SetConstF, inst.dest, out.AddConstant(gpr.GetImm(inst.src1)));
+			} else {
+				gpr.MapIn(inst.src1);
+				goto doDefault;
+			}
+			break;
+
+		case IROp::MfHi:
+		case IROp::MfLo:
+			gpr.MapDirty(inst.dest);
+			goto doDefault;
+
+		case IROp::MtHi:
+		case IROp::MtLo:
+			gpr.MapIn(inst.src1);
+			goto doDefault;
+
+		case IROp::Store8:
+		case IROp::Store16:
+		case IROp::Store32:
+			if (gpr.IsImm(inst.src1) && inst.src1 != inst.dest) {
+				gpr.MapIn(inst.dest);
+				out.Write(inst.op, inst.dest, 0, out.AddConstant(gpr.GetImm(inst.src1) + constants[inst.src2]));
+			} else {
+				gpr.MapInIn(inst.dest, inst.src1);
+				goto doDefault;
+			}
+			break;
+		case IROp::StoreFloat:
+			if (gpr.IsImm(inst.src1)) {
+				out.Write(inst.op, inst.dest, 0, out.AddConstant(gpr.GetImm(inst.src1) + constants[inst.src2]));
+			} else {
+				gpr.MapIn(inst.src1);
+				goto doDefault;
+			}
+			break;
+
+		case IROp::Load8:
+		case IROp::Load8Ext:
+		case IROp::Load16:
+		case IROp::Load16Ext:
+		case IROp::Load32:
+			if (gpr.IsImm(inst.src1) && inst.src1 != inst.dest) {
+				gpr.MapDirty(inst.dest);
+				out.Write(inst.op, inst.dest, 0, out.AddConstant(gpr.GetImm(inst.src1) + constants[inst.src2]));
+			} else {
+				gpr.MapDirtyIn(inst.dest, inst.src1);
+				goto doDefault;
+			}
+			break;
+		case IROp::LoadFloat:
+			if (gpr.IsImm(inst.src1)) {
+				out.Write(inst.op, inst.dest, 0, out.AddConstant(gpr.GetImm(inst.src1) + constants[inst.src2]));
+			} else {
+				gpr.MapIn(inst.src1);
+				goto doDefault;
+			}
+			break;
+
+		// FP-only instructions don't need to flush immediates.
+		case IROp::FAdd:
+		case IROp::FMul:
+		case IROp::FDiv:
+		case IROp::FSub:
+		case IROp::FNeg:
+		case IROp::FAbs:
+		case IROp::FSqrt:
+		case IROp::FMov:
+			out.Write(inst);
+			break;
+
+		case IROp::Syscall:
+		case IROp::Interpret:
+		case IROp::ExitToConst:
+		case IROp::ExitToReg:
+		case IROp::ExitToConstIfEq:
+		case IROp::ExitToConstIfNeq:
+		case IROp::ExitToConstIfFpFalse:
+		case IROp::ExitToConstIfFpTrue:
+		case IROp::ExitToConstIfGeZ:
+		case IROp::ExitToConstIfGtZ:
+		case IROp::ExitToConstIfLeZ:
+		case IROp::ExitToConstIfLtZ:
+		default:
+		{
+		doDefaultAndFlush:
+			gpr.FlushAll();
```

Comment:
Maybe we can have a temp that discards temps (like IRTEMP0) that are no longer read from due to other optimization passes (like const folding.)

-[Unknown]